File: Examples/SciChartDemo/SciChartDemo/Examples/Examples2D/CreateCustomCharts/SplineLineSeries/SplineLineRenderableSeries.m
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2019. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// SplineLineRenderableSeries.m is part of the SCICHART® Examples. Permission is hereby granted
// to modify, create derivative works, distribute and publish any part of this source
// code whether for commercial, private or personal use.
//
// The SCICHART® examples are distributed in the hope that they will be useful, but
// without any warranty. It is provided "AS IS" without warranty of any kind, either
// expressed or implied.
//******************************************************************************

#import "SplineLineRenderableSeries.h"
#import <SciChart/SCIXyRenderableSeriesBase+Protected.h>
#import "TriDiagonalMatrixF.h"
#import "CubicSpline.h"

@interface SplineLineRenderableSeries ()

@property (strong, nonatomic) SCISmartPropertyBool *isSplineEnabledProperty;
@property (strong, nonatomic) SCISmartPropertyInt *upSampleFactorProperty;

@end

@implementation SplineLineRenderableSeries {
    SCIFloatValues *_splineXCoords;
    SCIFloatValues *_splineYCoords;
}

- (instancetype)init {
    id<ISCIHitProvider> hitProvider = [[SCICompositeHitProvider alloc] initWithProvider1:[SCIPointMarkerHitProvider new] provider2:[SCILineHitProvider new]];
    self = [super initWithRenderPassData:[SCILineRenderPassData new] hitProvider:hitProvider nearestPointProvider:[SCINearestXyPointProvider new]];
    if (self) {
        _splineXCoords = [SCIFloatValues new];
        _splineYCoords = [SCIFloatValues new];
        
        _isSplineEnabledProperty = [[SCISmartPropertyBool alloc] initWithListener:self.boolInvalidateElementListener defaultValue:YES];
        _upSampleFactorProperty = [[SCISmartPropertyInt alloc] initWithListener:self.intInvalidateElementListener defaultValue:10];
    }
    return self;
}

- (BOOL)isSplineEnabled {
    return _isSplineEnabledProperty.value;
}

- (void)setIsSplineEnabled:(BOOL)isSplineEnabled {
    [_isSplineEnabledProperty setStrongValue:isSplineEnabled];
}

- (int)upSampleFactor {
    return _upSampleFactorProperty.value;
}

- (void)setUpSampleFactor:(int)upSampleFactor {
    [_upSampleFactorProperty setStrongValue:upSampleFactor];
}

- (void)disposeCachedData {
    [super disposeCachedData];
    
    [_splineXCoords dispose];
    [_splineYCoords dispose];
}

- (void)internalDrawWithContext:(id<ISCIRenderContext2D>)renderContext assetManager:(id<ISCIAssetManager2D>)assetManager renderPassData:(id<ISCISeriesRenderPassData>)renderPassData {
    // Don't draw transparent series
    float opacity = self.opacity;
    if (opacity == 0) return;
    
    SCIPenStyle *strokeStyle = self.strokeStyle;
    if (strokeStyle == nil || !strokeStyle.isVisible) return;
    
    SCILineRenderPassData *rpd = (SCILineRenderPassData *)renderPassData;
    id<ISCIDrawingContext> linesStripDrawingContext = SCIDrawingContextFactory.linesStripDrawingContext;
    
    id<ISCIPen2D> pen = [assetManager penWithStyle:strokeStyle andOpacity:opacity];
    [self computeSplineSeriesWithX:_splineXCoords y:_splineYCoords renderPassData:rpd isSplineEnabled:self.isSplineEnabled upSampleFactor:self.upSampleFactor];
    
    BOOL digitalLine = rpd.isDigitalLine;
    BOOL closeGaps = rpd.closeGaps;

    id<ISCISeriesDrawingManager> drawingManager = [self.services getServiceOfType:@protocol(ISCISeriesDrawingManager)];
    [drawingManager beginDrawWithContext:renderContext renderPassData:rpd];

    [drawingManager iterateLinesWith:linesStripDrawingContext pathColor:pen xCoords:_splineXCoords yCoords:_splineYCoords isDigitalLine:digitalLine closeGaps:closeGaps];

    [drawingManager endDraw];
    
    [self drawPointMarkersWithContext:renderContext assetManager:assetManager xCoords:rpd.xCoords yCoords:rpd.yCoords];
}

- (void)computeSplineSeriesWithX:(SCIFloatValues *)splineXCoords y:(SCIFloatValues *)splineYCoords renderPassData:(SCILineRenderPassData *)renderPassData isSplineEnabled:(BOOL)isSplineEnabled upSampleFactor:(int)upSampleFactor {
    if (!isSplineEnabled) return;
    
    // Spline enabled
    NSInteger count = self.currentRenderPassData.pointsCount;
    NSInteger splineCount = count * upSampleFactor;
    
    splineXCoords.count = splineCount;
    splineYCoords.count = splineCount;
    
    SCIFloatValues *xs = [[SCIFloatValues alloc] initWithCapacity:splineCount];
    float *xData = renderPassData.xCoords.itemsArray;
    float stepSize = (xData[count - 1] - xData[0]) / (splineCount - 1);
    
    // set spline xCoords
    for (int i = 0; i < splineCount; i++) {
        float value = xData[0] + i * stepSize;
        [xs add:value];
    }
    
    CubicSpline *cubicSpline = [CubicSpline new];
    SCIFloatValues *ys = [cubicSpline fitAndEvalX:renderPassData.xCoords y:renderPassData.yCoords xS:xs startSlope:NAN endSlope:NAN];
    
    [splineXCoords clear];
    [splineYCoords clear];
    [splineXCoords addValues:xs.itemsArray count:xs.count];
    [splineYCoords addValues:ys.itemsArray count:ys.count];
}

@end

File: Examples/SciChartDemo/SciChartDemo/Examples/Examples2D/CreateCustomCharts/SplineLineSeries/SplineLineRenderableSeries.swift
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2019. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// SplineLineRenderableSeries.swift is part of the SCICHART® Examples. Permission is hereby granted
// to modify, create derivative works, distribute and publish any part of this source
// code whether for commercial, private or personal use.
//
// The SCICHART® examples are distributed in the hope that they will be useful, but
// without any warranty. It is provided "AS IS" without warranty of any kind, either
// expressed or implied.
//******************************************************************************

import SciChart.Protected.SCIXyRenderableSeriesBase

class SplineLineRenderableSeries: SCIXyRenderableSeriesBase {
    lazy var isSplineEnabledProperty = SCISmartPropertyBool(listener: boolInvalidateElementListener, defaultValue: true)
    lazy var upSampleFactorProperty = SCISmartPropertyInt(listener: intInvalidateElementListener, defaultValue: 10)
    
    let splineXCoords = SCIFloatValues()
    let splineYCoords = SCIFloatValues()
    
    override init() {
        let hitProvider = SCICompositeHitProvider(provider1: SCIPointMarkerHitProvider(), provider2: SCILineHitProvider())
        super.init(renderPassData: SCILineRenderPassData(), hitProvider: hitProvider, nearestPointProvider: SCINearestXyPointProvider())
    }
    
    var isSplineEnabled: Bool {
        get { return isSplineEnabledProperty.value }
        set(value) { isSplineEnabledProperty.setStrongValue(value) }
    }
    
    var upSampleFactor : Int {
        get { return Int(upSampleFactorProperty.value) }
        set(value) { upSampleFactorProperty.setStrongValue(Int32(value)) }
    }
    
    override func disposeCachedData() {
        super.disposeCachedData()
        
        splineXCoords.dispose()
        splineYCoords.dispose()
    }

    override func internalDraw(with renderContext: ISCIRenderContext2D!, assetManager: ISCIAssetManager2D!, renderPassData: ISCISeriesRenderPassData!) {
        // Don't draw transparent series
        guard self.opacity == 0 else { return }
        guard let strokeStyle = self.strokeStyle else { return }
        guard strokeStyle.isVisible else { return }
        
        let rpd = renderPassData as! SCILineRenderPassData
        let linesStripDrawingContext = SCIDrawingContextFactory.linesStripDrawingContext
        
        let pen = assetManager.pen(with: strokeStyle, andOpacity: opacity)
        computeSplineSeries(splineXCoords: splineXCoords, splineYCoords: splineYCoords, renderPassData: rpd, isSplineEnabled: self.isSplineEnabled, upSampleFactor: self.upSampleFactor)
        
        let digitalLine = rpd.isDigitalLine
        let closeGaps = rpd.closeGaps
        
        let drawingManager = services.getServiceOfType(ISCISeriesDrawingManager.self) as! ISCISeriesDrawingManager
        drawingManager.beginDraw(with: renderContext, renderPassData: rpd)
        
        drawingManager.iterateLines(with: linesStripDrawingContext, pathColor: pen, xCoords: splineXCoords, yCoords: splineYCoords, isDigitalLine: digitalLine, closeGaps: closeGaps)
        
        drawingManager.endDraw()
        
        drawPointMarkers(with: renderContext, assetManager: assetManager, xCoords: rpd.xCoords, yCoords: rpd.yCoords)
    }
    
    fileprivate func computeSplineSeries(splineXCoords: SCIFloatValues, splineYCoords: SCIFloatValues, renderPassData: SCILineRenderPassData, isSplineEnabled: Bool, upSampleFactor: Int) {
        if (!isSplineEnabled) { return }
        
        let count = currentRenderPassData.pointsCount
        let splineCount = count * upSampleFactor
        
        splineXCoords.count = splineCount
        splineYCoords.count = splineCount
        
        let x = renderPassData.xCoords.itemsArray
        let y = renderPassData.yCoords.itemsArray
        
        var xs = splineXCoords.itemsArray
        let stepSize = (x[x.count - 1] - x[0]) / Float(splineCount - 1)
        
        // set spline xCoords
        for i in 0..<splineCount {
            xs[Int(i)] = x[0] + Float(i) * stepSize
        }
        
        let cubicSpline = CubicSpline()
        let ys = cubicSpline.fitAndEval(x: x, y: y, xs: xs, startSlope: Float.nan, endSlope: Float.nan)
        
        splineXCoords.clear()
        splineYCoords.clear()
        splineXCoords.add(values: xs)
        splineYCoords.add(values: ys.itemsArray)
    }
}
